### DISCUSSION TITLE
How to use Dynamic Form Data in FastAPI application function Python?

### DISCUSSION THREAD

**Discussion by RakshithRAcharya** [fastapi/fastapi#8745] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/questions/68375506/how-to-use-dynamic-form-data-in-fastapi-application-function-python) with my use case:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
def read_root():
    return {"Hello": "World"}

@app.post("/do_something")
def do_something(data:str = Form(...)): # Here this data: str=Form(...)should be dynamic.....taking data from HTML form
   print(data)
   return ""

```

### Description

Normally If we have form data like:
```
<form action="/do_something" method="post">
   <label>SomeLabel</label>
   <input type="text" name="data">
   <input type="submit" value="Submit">
</form>
```
We can use something like this in FastAPI application
```
@app.post("/do_something")
def do_something(data:str = Form(...)):
   print(data)
   return ""
```
But when the HTML changes according to inputs like:
```
<form action="/do_something" method="post">
   {%for i in some_list%}
   <label>{{i}}</label>
   <input type="text" name="{{i}}">
   {%endfor%}
   <input type="submit" value="Submit">
</form>
```
So then what do we do for taking all these inputs as parameters for the function. Also the /docs of the FastAPI application should contain these things as text fields. Something like this: ![Take a look at what it should look like on /docs](https://i.stack.imgur.com/qJo1J.png)

Please Help Fast... Thanks

Also the known solution is something around using request and request.form()
```
@app.post("/do_something")
def do_something(request: Request):
   data = request.form()
   print(data)
   return ""
```
But in this approach nothing shows up on /docs endpoint. I want the data to show up on the swagger docs also

### Environment

* OS: [e.g. Linux / Windows / macOS]: macOS
* FastAPI Version [e.g. 0.3.0]: 0.65.2
* Python version: 3.8.10

**Comment by boanergies** [comment]
Could you maybe instead receive a List of Forms and then parse the data from there? It feels like your asking for the route in fastapi to dynamically react to what it is given as part of the jinja2 for loop which doesn't seem possible from what I can see.

**Comment by RakshithRAcharya** [comment]
> Could you maybe instead receive a List of Forms and then parse the data from there? It feels like your asking for the route in fastapi to dynamically react to what it is given as part of the jinja2 for loop which doesn't seem possible from what I can see.

Could you please give a snippet on accepting a list of forms. Thank you

**Comment by boanergies** [comment]
When I think about this more its not realistic to expect a 'List' of Forms.

Rather it would be important for the forms your building to extract what is constant and what is dynamic and have the route work with the data as it is updated.

Like if someone wanted a list of aliases along with a username that could be updated
```Python
@app.post("/do_something")
def do_something(username: str = Form(...), aliases: List[str] = Form(...)):
```

If your use case is literally having an entire Form be dynamic you may have to parse the request directly like your potential solution shows. The idea of Forms from what I understand is its not expected the entire data to be parsed can not be known before doing so.

**Comment by RakshithRAcharya** [comment]
> When I think about this more its not realistic to expect a 'List' of Forms.
> 
> Rather it would be important for the forms your building to extract what is constant and what is dynamic and have the route work with the data as it is updated.
> 
> Like if someone wanted a list of aliases along with a username that could be updated
> 
> ```python
> @app.post("/do_something")
> def do_something(username: str = Form(...), aliases: List[str] = Form(...)):
> ```
> 
> If your use case is literally having an entire Form be dynamic you may have to parse the request directly like your potential solution shows. The idea of Forms from what I understand is its not expected the entire data to be parsed can not be known before doing so.

So then I cannot parse the form dynamically.......So can I close this issue/question then??......Or you could go ahead and close it

**Comment by boanergies** [comment]
I'm just a fellow user who is answering questions because I'm waiting for my own to be answered. 

Whether to close the issue is up to you. If you feel your question has been answered, then close the issue, otherwise you can wait to see if someone else gives a different answer.

**Comment by YuriiMotov** [comment, marked as answer]
Make it fully dynamic seems to be impossible.
But If you just want to keep the form and the processing endpoint in sync, then something like this should help:

```py
from typing import Annotated

from fastapi import FastAPI, Form, Request
from fastapi.templating import Jinja2Templates
from pydantic import BaseModel


templates = Jinja2Templates(directory="templates")


class FormParams(BaseModel):
    name: str
    age: int
    other: str


app = FastAPI()


@app.get("/")
async def get_form(request: Request):
    return templates.TemplateResponse(
        request=request,
        name="form.html",
        context={"fields": list(FormParams.model_fields.keys())},
    )


@app.post("/")
async def post_form(form: Annotated[FormParams, Form()]):
    return form
```

Template (`templates/form.html`):
```html
<form action="/do_something" method="post">
   {%for i in fields%}
   <label>{{i}}</label>
   <input type="text" name="{{i}}">
   {%endfor%}
   <input type="submit" value="Submit">
</form>
```

Whenever you update `FormParams` and reload the app - your html form and OpenAPI schema will be in sync

---

> Please Help Fast... Thanks

Hope it's not too late :)
